File: frontend/src/app/core/services/cart.service.ts
import { Injectable, inject, signal, computed } from '@angular/core';
import { Observable, tap, catchError, of } from 'rxjs';
import { ApiService } from './api.service';

/**
 * Interfaces para el carrito
 */
export interface CartItem {
  id: number;
  productId: number;
  productName: string;
  productImage: string;
  productPrice: number;
  quantity: number;
  subtotal: number;
}

export interface Cart {
  items: CartItem[];
  totalItems: number;
  totalPrice: number;
}

export interface AddToCartRequest {
  productId: number;
  quantity: number;
}

/**
 * CartService - Servicio del Carrito de Compras
 *
 * Gestiona el carrito de compras del usuario.
 * Se sincroniza con el backend cuando el usuario está autenticado.
 */
@Injectable({
  providedIn: 'root',
})
export class CartService {
  private readonly api = inject(ApiService);

  // Estado reactivo
  private readonly _cart = signal<Cart>({
    items: [],
    totalItems: 0,
    totalPrice: 0,
  });
  private readonly _isLoading = signal(false);

  /** Carrito actual */
  readonly cart = this._cart.asReadonly();

  /** Estado de carga */
  readonly isLoading = this._isLoading.asReadonly();

  /** Items del carrito */
  readonly items = computed(() => this._cart().items);

  /** Número total de items */
  readonly totalItems = computed(() => this._cart().totalItems);

  /** Precio total */
  readonly totalPrice = computed(() => this._cart().totalPrice);

  /** ¿El carrito está vacío? */
  readonly isEmpty = computed(() => this._cart().items.length === 0);

  // ═══════════════════════════════════════════════════════════════
  // MÉTODOS DEL CARRITO
  // ═══════════════════════════════════════════════════════════════

  /**
   * Carga el carrito desde el backend.
   */
  loadCart(): Observable<{ success: boolean; data: Cart }> {
    this._isLoading.set(true);

    return this.api.get<{ success: boolean; data: Cart }>('/cart').pipe(
      tap((response) => {
        if (response.success) {
          this._cart.set(response.data);
        }
        this._isLoading.set(false);
      }),
      catchError((error) => {
        this._isLoading.set(false);
        console.error('[CartService] Error cargando carrito:', error);
        return of({ success: false, data: { items: [], totalItems: 0, totalPrice: 0 } });
      })
    );
  }

  /**
   * Añade un producto al carrito.
   */
  addToCart(productId: number, quantity: number = 1): Observable<{ success: boolean; data: Cart }> {
    this._isLoading.set(true);

    const request: AddToCartRequest = { productId, quantity };

    return this.api.post<{ success: boolean; data: Cart }>('/cart/items', request).pipe(
      tap((response) => {
        if (response.success) {
          this._cart.set(response.data);
          console.log('[CartService] Producto añadido al carrito');
        }
        this._isLoading.set(false);
      }),
      catchError((error) => {
        this._isLoading.set(false);
        console.error('[CartService] Error añadiendo al carrito:', error);
        throw error;
      })
    );
  }

  /**
   * Actualiza la cantidad de un producto.
   */
  updateQuantity(productId: number, quantity: number): Observable<{ success: boolean; data: Cart }> {
    this._isLoading.set(true);

    return this.api.put<{ success: boolean; data: Cart }>(`/cart/items/${productId}?quantity=${quantity}`, {}).pipe(
      tap((response) => {
        if (response.success) {
          this._cart.set(response.data);
        }
        this._isLoading.set(false);
      }),
      catchError((error) => {
        this._isLoading.set(false);
        console.error('[CartService] Error actualizando cantidad:', error);
        throw error;
      })
    );
  }

  /**
   * Elimina un producto del carrito.
   */
  removeFromCart(productId: number): Observable<{ success: boolean; data: Cart }> {
    this._isLoading.set(true);

    return this.api.delete<{ success: boolean; data: Cart }>(`/cart/items/${productId}`).pipe(
      tap((response) => {
        if (response.success) {
          this._cart.set(response.data);
          console.log('[CartService] Producto eliminado del carrito');
        }
        this._isLoading.set(false);
      }),
      catchError((error) => {
        this._isLoading.set(false);
        console.error('[CartService] Error eliminando del carrito:', error);
        throw error;
      })
    );
  }

  /**
   * Vacía el carrito completo.
   */
  clearCart(): Observable<{ success: boolean; data: Cart }> {
    this._isLoading.set(true);

    return this.api.delete<{ success: boolean; data: Cart }>('/cart').pipe(
      tap((response) => {
        if (response.success) {
          this._cart.set(response.data);
          console.log('[CartService] Carrito vaciado');
        }
        this._isLoading.set(false);
      }),
      catchError((error) => {
        this._isLoading.set(false);
        console.error('[CartService] Error vaciando carrito:', error);
        throw error;
      })
    );
  }

  /**
   * Procesa el checkout (simulado).
   */
  checkout(): Observable<{ success: boolean; data: string; message: string }> {
    this._isLoading.set(true);

    return this.api.post<{ success: boolean; data: string; message: string }>('/cart/checkout', {}).pipe(
      tap((response) => {
        if (response.success) {
          // Limpiar el carrito local
          this._cart.set({
            items: [],
            totalItems: 0,
            totalPrice: 0,
          });
          console.log('[CartService] Checkout completado:', response.message);
        }
        this._isLoading.set(false);
      }),
      catchError((error) => {
        this._isLoading.set(false);
        console.error('[CartService] Error en checkout:', error);
        throw error;
      })
    );
  }

  /**
   * Incrementa la cantidad de un producto.
   */
  incrementQuantity(productId: number): void {
    const item = this._cart().items.find((i) => i.productId === productId);
    if (item) {
      this.updateQuantity(productId, item.quantity + 1).subscribe();
    }
  }

  /**
   * Decrementa la cantidad de un producto.
   */
  decrementQuantity(productId: number): void {
    const item = this._cart().items.find((i) => i.productId === productId);
    if (item && item.quantity > 1) {
      this.updateQuantity(productId, item.quantity - 1).subscribe();
    } else if (item && item.quantity === 1) {
      this.removeFromCart(productId).subscribe();
    }
  }
}

File: backend/src/main/java/com/educadino/service/CartService.java
package com.educadino.service;

import com.educadino.dto.*;
import com.educadino.entity.CartItem;
import com.educadino.entity.Product;
import com.educadino.entity.User;
import com.educadino.exception.BadRequestException;
import com.educadino.exception.ResourceNotFoundException;
import com.educadino.repository.CartItemRepository;
import com.educadino.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Servicio del Carrito de Compras
 *
 * Maneja las operaciones del carrito (simulado).
 */
@Service
@RequiredArgsConstructor
public class CartService {

    private final CartItemRepository cartItemRepository;
    private final ProductRepository productRepository;
    private final UserService userService;

    /**
     * Obtiene el carrito del usuario actual.
     */
    @Transactional(readOnly = true)
    public CartDto getCart() {
        User user = userService.getCurrentUser();
        List<CartItem> items = cartItemRepository.findByUserId(user.getId());

        List<CartItemDto> itemDtos = items.stream()
                .map(this::mapToDto)
                .collect(Collectors.toList());

        BigDecimal total = items.stream()
                .map(item -> item.getProduct().getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        int totalItems = items.stream()
                .mapToInt(CartItem::getQuantity)
                .sum();

        return CartDto.builder()
                .items(itemDtos)
                .totalItems(totalItems)
                .totalPrice(total)
                .build();
    }

    /**
     * Añade un producto al carrito.
     */
    @Transactional
    public CartDto addToCart(AddToCartRequest request) {
        User user = userService.getCurrentUser();

        Product product = productRepository.findById(request.getProductId())
                .orElseThrow(() -> new ResourceNotFoundException("Producto", request.getProductId()));

        // Verificar stock
        if (product.getStock() < request.getQuantity()) {
            throw new BadRequestException("No hay suficiente stock disponible");
        }

        // Buscar si ya existe el producto en el carrito
        var existingItem = cartItemRepository.findByUserIdAndProductId(user.getId(), product.getId());

        if (existingItem.isPresent()) {
            // Actualizar cantidad
            CartItem item = existingItem.get();
            int newQuantity = item.getQuantity() + request.getQuantity();

            if (product.getStock() < newQuantity) {
                throw new BadRequestException("No hay suficiente stock disponible");
            }

            item.setQuantity(newQuantity);
            cartItemRepository.save(item);
        } else {
            // Crear nuevo item
            CartItem item = CartItem.builder()
                    .user(user)
                    .product(product)
                    .quantity(request.getQuantity())
                    .build();
            cartItemRepository.save(item);
        }

        return getCart();
    }

    /**
     * Actualiza la cantidad de un producto en el carrito.
     */
    @Transactional
    public CartDto updateQuantity(Long productId, int quantity) {
        User user = userService.getCurrentUser();

        if (quantity <= 0) {
            return removeFromCart(productId);
        }

        CartItem item = cartItemRepository.findByUserIdAndProductId(user.getId(), productId)
                .orElseThrow(() -> new ResourceNotFoundException("Item no encontrado en el carrito"));

        if (item.getProduct().getStock() < quantity) {
            throw new BadRequestException("No hay suficiente stock disponible");
        }

        item.setQuantity(quantity);
        cartItemRepository.save(item);

        return getCart();
    }

    /**
     * Elimina un producto del carrito.
     */
    @Transactional
    public CartDto removeFromCart(Long productId) {
        User user = userService.getCurrentUser();
        cartItemRepository.deleteByUserIdAndProductId(user.getId(), productId);
        return getCart();
    }

    /**
     * Vacía el carrito completo.
     */
    @Transactional
    public CartDto clearCart() {
        User user = userService.getCurrentUser();
        cartItemRepository.deleteByUserId(user.getId());
        return CartDto.builder()
                .items(List.of())
                .totalItems(0)
                .totalPrice(BigDecimal.ZERO)
                .build();
    }

    /**
     * Simula el checkout del carrito.
     */
    @Transactional
    public ApiResponse<String> checkout() {
        User user = userService.getCurrentUser();
        List<CartItem> items = cartItemRepository.findByUserId(user.getId());

        if (items.isEmpty()) {
            throw new BadRequestException("El carrito está vacío");
        }

        // Simular la compra (en una app real, aquí iría el procesamiento de pago)
        BigDecimal total = items.stream()
                .map(item -> item.getProduct().getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        // Vaciar el carrito
        cartItemRepository.deleteByUserId(user.getId());

        return ApiResponse.success(
                "Pedido completado",
                "¡Compra simulada exitosa! Total: €" + total
        );
    }

    /**
     * Convierte CartItem a CartItemDto.
     */
    private CartItemDto mapToDto(CartItem item) {
        return CartItemDto.builder()
                .id(item.getId())
                .productId(item.getProduct().getId())
                .productName(item.getProduct().getName())
                .productImage(item.getProduct().getImage())
                .productPrice(item.getProduct().getPrice())
                .quantity(item.getQuantity())
                .subtotal(item.getProduct().getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
                .build();
    }
}
